diff -rupN generated/nodes-reputation.c generated-rep/nodes-reputation.c
--- generated/nodes-reputation.c	2014-05-08 09:28:06.000000000 +0200
+++ generated-rep/nodes-reputation.c	2014-05-08 09:27:02.000000000 +0200
@@ -1,4 +1,5 @@
 #include "nodes-reputation.h"
+#include "../lib/log.h"
 void anonymous5(node_t* from, node_t* hop, node_t* to, payload_t* payload) {
   if((hop == to)) {
     return;
@@ -6,6 +7,9 @@ void anonymous5(node_t* from, node_t* ho
   if((hop->address == 0)) {
     return;
   }
+  _log("RP: tracking payload from %02x %02x to %02x %02x : size=%d\n",
+       (uint8_t)(hop->address >> 8), (uint8_t)hop->address,
+       (uint8_t)(to->address  >> 8), (uint8_t)to->address, payload->size);
   hop->msg_count++;
   list_of_tuple_0_ts_push(&hop->queue, make_tuple_0_t((now() + forward_timeout),
                           payload));
@@ -15,14 +19,29 @@ void anonymous7(node_t* me, node_t* send
   if((from != me)) {
     return;
   }
-  list_of_tuple_0_ts_remove_match_arg_0(&sender->queue, payload);
+  if( list_of_tuple_0_ts_remove_match_arg_0(&sender->queue, payload) > 0 ) {
+    _log("RP: cleared payload from %02x %02x : size=%d\n",
+         (uint8_t)(from->address >> 8), (uint8_t)from->address, payload->size);
+  } else {
+    _log("RP: unexpected payload from %02x %02x\n",
+         (uint8_t)(from->address >> 8), (uint8_t)from->address);
+  }
 }
 void anonymous12(node_t* node) {
   uint16_t failures = list_of_tuple_0_ts_remove_match_lt_now(&node->queue);
+  if(failures>0) {
+    _log("RP: late: %02x %02x\n",
+         (uint8_t)(node->address >> 8), (uint8_t)node->address);
+  }
+  _log("RP: validating node %02x %02x : fail=%d/%d : a=%.2f b=%.2f t=%.2f\n",
+       (uint8_t)(node->address >> 8), (uint8_t)node->address,
+       failures, node->msg_count,
+       (double)node->alpha, (double)node->beta, (double)node->trust);
   node->alpha = (((aging_weight * node->alpha) + node->msg_count) - failures);
   node->beta = ((aging_weight * node->beta) + failures);
   node->trust = ((node->alpha + 1) / ((node->alpha + node->beta) + 2));
   if((node->trust < 0.25)) {
+    _log("RP: trust lost\n");
     nodes_send(node, 2, 0x00, 0x02);
   }
   node->msg_count = 0;
